linking.go

```diff
@@ -2,147 +2,138 @@ package ipld
-//
-// LinkContext objects can be provided to give additional information
-// to the loader, and will be automatically filled out when a Loader
-// is used by systems in the traversal package; most Loader implementations
-// should also work fine when given the zero value of LinkContext.
-//
-// Loaders are implicitly coupled to a Link implementation and have some
-// "extra" knowledge of the concrete Link type.  This necessary since there is
-// no mandated standard for how to serially represent Link itself, and such
-// a representation is typically needed by a Storer implementation.
-type Loader func(lnk Link, lnkCtx LinkContext) (io.Reader, error)
+func (ls *LinkSystem) MustLoad(lnkCtx LinkContext, lnk Link, np NodePrototype) Node {
+	if n, err := ls.Load(lnkCtx, lnk, np); err != nil {
+		panic(err)
+	} else {
+		return n
+	}
+}
 
-// Storer functions are used to a get a writer for raw serialized content,
-// which will be committed to storage indexed by Link.
-// A stoerer function is used by providing it to a LinkBuilder.Build() call.
-//
-// The storer system comes in two parts: the Storer itself *starts* a storage
-// operation (presumably to some e.g. tempfile) and returns a writer; the
-// StoreCommitter returned with the writer is used to *commit* the final storage
-// (much like a 'Close' operation for the writer).
-//
-// Storers typically have some filesystem or database handle contained
-// within their closure which is used to satisfy read operations.
-//
-// LinkContext objects can be provided to give additional information
-// to the storer, and will be automatically filled out when a Storer
-// is used by systems in the traversal package; most Storer implementations
-// should also work fine when given the zero value of LinkContext.
-//
-// Storers are implicitly coupled to a Link implementation and have some
-// "extra" knowledge of the concrete Link type.  This necessary since there is
-// no mandated standard for how to serially represent Link itself, and such
-// a representation is typically needed by a Storer implementation.
-type Storer func(lnkCtx LinkContext) (io.Writer, StoreCommitter, error)
+func (ls *LinkSystem) Fill(lnkCtx LinkContext, lnk Link, na NodeAssembler) error {
+	if lnkCtx.Ctx == nil {
+		lnkCtx.Ctx = context.Background()
+	}
+	// Choose all the parts.
+	decoder, err := ls.DecoderChooser(lnk)
+	if err != nil {
+		return ErrLinkingSetup{"could not choose a decoder", err}
+	}
+	hasher, err := ls.HasherChooser(lnk.Prototype())
+	if err != nil {
+		return ErrLinkingSetup{"could not choose a hasher", err}
+	}
+	if ls.StorageReadOpener == nil {
+		return ErrLinkingSetup{"no storage configured for reading", io.ErrClosedPipe} // REVIEW: better cause?
+	}
+	// Open storage, read it, verify it, and feed the codec to assemble the nodes.
+	reader, err := ls.StorageReadOpener(lnkCtx, lnk)
+	if err != nil {
+		return err
+	}
+	tee := io.TeeReader(reader, hasher)
```

Comment:
@hannahhoward you might wanna be aware: I didn't carry forward the byteAccessor feature detector.  That might be a "yet" thing -- do you know if we'll be in a good place to detect a performance regression in the area that made us introduce this the first time?

I'm weakly betting it might less relevant now because the hasher doesn't force excessive copies after this rework, but, it's only a bet.

Reply:
It's back in.  Reintroduced with 08f13e569dc4264773e99657a84319952d944a70.